crates/red_knot/src/types.rs

```diff
@@ -46,6 +51,45 @@ impl Type {
     pub const fn is_unknown(&self) -> bool {
         matches!(self, Type::Unknown)
     }
+
+    // NOTE: naming? `get_member_type`
+    pub fn get_member(&self, db: &dyn SemanticDb, name: &Name) -> QueryResult<Type> {
+        // NOTE: maybe make this a trait that each TypeID implements?
+        match self {
+            Type::Any => todo!("attribute lookup on Any type"),
+            Type::Never => todo!("attribute lookup on Never type"),
+            Type::Unknown => todo!("attribute lookup on Unknown type"),
+            Type::Unbound => todo!("attribute lookup on Unbound type"),
+            Type::Function(_) => todo!("attribute lookup on Function type"),
+            Type::Module(module_id) => module_id
+                .get_member(db, name)
+                // FIXME: don't want to unwrap here; need to add a QueryError ctor for the case
+                // where the get_member result is Ok(None)
+                .map(Option::unwrap),
+            Type::Class(class_id) => class_id
+                .get_own_class_member(db, name)
+                .or_else(|_| class_id.get_super_class_member(db, name))
+                // FIXME: don't want to unwrap here; need to add a QueryError ctor for the case
+                // where the get_member result is Ok(None)
+                .map(Option::unwrap),
+            Type::Instance(_) => {
+                // TODO MRO? get_own_instance_member, get_instance_member
+                todo!("attribute lookup on Instance type")
+            }
+            Type::Union(union_id) => {
+                let jar: &SemanticJar = db.jar()?;
+                let union_ref = jar.type_store.get_union(*union_id);
+                // TODO return a type FF at least one of the unioned-types have the member; return
+                // the union over those member types
+                todo!("attribute lookup on Union type")
+            }
+            Type::Intersection(_) => {
+                // TODO return a type IFF all of the intersected-types have the member; but what
+                // type?
```

Comment:
In an intersection, the type must be all of the intersected types, so if any of those have the member, it has the member. So we can return a type if any of the types have it; if more than one has it, we should return the intersection of those types.

Reply:
* No types have the member, return `None`
* One type has the member, return the member type (or a singleton intersection containing it?).
* Two types have the member, return the intersection of the member types.
* Two types have the member, but one type doesn't have the member...? It seems to me that the intersection ought to be empty.

Reply:
There never a need for singleton intersection (or union), it should  always simplify to just the single type. 

In the final case, the type that is missing the attribute should just contribute nothing to the intersection, it shouldn't contribute the bottom type (thus making the intersection empty). I would have to think a bit more about the right way to think about this formally, but I'm sure it's the right behavior. A & B & C must be an A and a B and a C, so if only A and B have the attribute, we know the attribute is present and its type is A.foo & B.foo; C contributes no information about its type.